transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringBufferRing.java

```diff
@@ -0,0 +1,452 @@
+
+    private BufferRingExhaustedEvent exhaustedEvent;
+
+    IoUringBufferRing(int ringFd, long ioUringBufRingAddr,
+                      short entries, short bufferGroupId,
+                      int chunkSize, IoUringIoHandler ioUringIoHandler,
+                      ByteBufAllocator byteBufAllocator
+    ) {
+        this.ioUringBufRingAddr = ioUringBufRingAddr;
+        this.entries = entries;
+        this.mask = (short) (entries - 1);
+        this.bufferGroupId = bufferGroupId;
+        this.ringFd = ringFd;
+        this.userspaceBufferHolder = new ByteBuf[entries];
+        this.nextIndex = 0;
+        this.chunkSize = chunkSize;
+        this.hasSpareBuffer = false;
+        this.source = ioUringIoHandler;
+        this.byteBufAllocator = byteBufAllocator;
+        this.recycleBufferTasks = new Runnable[entries];
+    }
+
+    void markReadFail() {
+        hasSpareBuffer = false;
+    }
+
+    boolean hasSpareBuffer() {
+        return hasSpareBuffer;
+    }
+
+    /**
+     * lazy get the event, if the event is null, create a new one
+     * @return a BufferRingExhaustedEvent Instance
+     */
+    BufferRingExhaustedEvent getExhaustedEvent() {
+        BufferRingExhaustedEvent res = exhaustedEvent;
+        if (res == null) {
+            res = new BufferRingExhaustedEvent(bufferGroupId);
+            exhaustedEvent = res;
+        }
+        return res;
+    }
+
+    void recycleBuffer(short bid) {
+        source.submitBeforeIO(recycleBufferTasks[bid]);
+    }
+
+    void addToRing(short bid, boolean needAdvance) {
+        ByteBuf byteBuf = userspaceBufferHolder[bid];
+        long tailFieldAddress = ioUringBufRingAddr + Native.IO_URING_BUFFER_RING_TAIL;
+        short oldTail = PlatformDependent.getShort(tailFieldAddress);
+        int ringIndex = oldTail & mask;
+        //  see:
+        //  https://github.com/axboe/liburing/blob/19134a8fffd406b22595a5813a3e319c19630ac9/src/include/liburing.h#L1561
+        long ioUringBufAddress = ioUringBufRingAddr + (long) Native.SIZEOF_IOURING_BUF * ringIndex;
+        PlatformDependent.putLong(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_ADDR, byteBuf.memoryAddress());
+        PlatformDependent.putInt(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_LEN, (short) byteBuf.capacity());
+        PlatformDependent.putShort(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_BID, bid);
+        if (needAdvance) {
+            advanceTail(1);
+        }
+    }
+
+    void appendBuffer(int count) {
+        int expectedIndex = nextIndex + count;
+        if (expectedIndex > entries) {
+            throw new IllegalStateException(
+                    String.format(
+                            "We want append %d buffer, but buffer ring is full. The ring hold %s buffers",
+                            count, entries)
+            );
+        }
+
+        for (int i = 0; i < count; i++) {
+            ByteBuf buffer = byteBufAllocator.ioBuffer(chunkSize);
```

Comment:
do we want to add them like this, allocating one  by one, or just have a single allocation and slice it in N parts?
It can be bothersome, but will reduce a bit the fragmentation

Reply:
Yeah, what you said makes sense. However, currently we will only apply lazily one by one. Maybe I can provide an additional initialization parameter to allow the user to pre-fill n buffers, so that the optimization you mentioned can be used.

Reply:
Also if the allocation is too big we might need pool it in the allocator.